File: carDemo/web/src/experiment/ai/geneticAlgorithm.js

import { Logger } from "../utils/index.js";

import ArtificialNeuralNetwork from "./artificialNeuralNetwork.js";

class GeneticAlgorithm {

	constructor(genomeSize, ann_topology) {

		this._logger = new Logger("logger");

		this._currentId = 0;

		this._ANNs = [];
		for (let ii = 0; ii < genomeSize; ++ii)
			this._ANNs.push(new ArtificialNeuralNetwork(ann_topology));

		this._genomes = [];
		for (let ii = 0; ii < genomeSize; ++ii)
			this._genomes.push({
				id: this._currentId++,
				fitness: 0.0,
				weights: this._ANNs[ii].weights
			});

		this._currentGeneration = 1;
		this._bestFitness = 0.0;
		this._bestGenome = null;
	}

	breedPopulation() {

		let progressWasMade = false;

		const bestGenomes = this._getBestGenomes(6);

		const children = [];

		if (this._bestFitness < bestGenomes[0].fitness)
			this._bestFitness = bestGenomes[0].fitness;

		if (this._bestGenome === null || this._bestGenome.fitness < this._bestFitness) {

			this._bestGenome = bestGenomes[0];
			progressWasMade = true;
		}

		{ // elitism

			const bestDude = {
				id: this._bestGenome.id,
				fitness: 0.0,
				weights: []
			};

			const weights = this._bestGenome.weights;
			for (let ii = 0; ii < weights.length; ++ii)
				bestDude.weights.push(weights[ii]);

			this._mutate(bestDude);

			children.push(bestDude);

		} // elitism

		// breed best genomes

		for (let ii = 0; ii < bestGenomes.length; ++ii)
			for (let jj = ii + 1; jj < bestGenomes.length; ++jj) {

				const	baby1 = {},
						baby2 = {};

				this._crossBreed(bestGenomes[ii], bestGenomes[jj], baby1, baby2);
				this._mutate(baby1);
				this._mutate(baby2);

				if (children.length < this._genomes.length)
					children.push(baby1);

				if (children.length < this._genomes.length)
					children.push(baby2);
			}

		// For the remainding n population, add some random kiddies.
		const remainingChildren = (this._genomes.length - children.length);

		const total_weights = this._ANNs[0].totalWeights;

		for (let ii = 0; ii < remainingChildren; ++ii) {

			const genome = {
				id: this._currentId++,
				fitness: 0.0,
				weights: []
			};

			for (let jj = 0; jj < total_weights; ++jj)
				genome.weights.push(Math.random() - Math.random());  // [-1 < x < +1]

			children.push(genome);
		}

		this._genomes = children;
		++this._currentGeneration;

		for (let ii = 0; ii < this._genomes.length; ++ii)
			this._ANNs[ii].weights = this._genomes[ii].weights;

		return progressWasMade;
	}

	_getBestGenomes(totalAsked) {

		// realistic total outputed genomes
		totalAsked = Math.min(this._genomes.length, totalAsked);

		const sortedGenomesIndex = [];
		for (let ii = 0; ii < this._genomes.length; ++ii)
			sortedGenomesIndex.push({ index: ii, fitness: this._genomes[ii].fitness });

		sortedGenomesIndex.sort((itemA, itemB) =>  itemB.fitness - itemA.fitness);

		const out = [];

		for (let ii = 0; ii < totalAsked; ++ii) {

			const index = sortedGenomesIndex[ii].index;

			out.push(this._genomes[index]);
		}

		return out;
	}

	_crossBreed(g1, g2, baby1, baby2) {

		const totalWeights = g1.weights.length;
		const crossover = Math.floor(Math.random() * totalWeights);

		baby1.id = this._currentId++;
		baby1.weights = [];

		baby2.id = this._currentId++;
		baby2.weights = [];

		for (let ii = 0; ii < crossover; ++ii) {
			baby1.weights[ii] = g1.weights[ii];
			baby2.weights[ii] = g2.weights[ii];
		}

		for (let ii = crossover; ii < totalWeights; ii++) {
			baby1.weights[ii] = g2.weights[ii];
			baby2.weights[ii] = g1.weights[ii];
		}
	}

	_mutate(genome) {

		for (let ii = 0; ii < genome.weights.length; ++ii)
			if (Math.random() < 0.1) {
				const randomValue = Math.random() - Math.random(); // [-1 < x < +1]
				genome.weights[ii] += (randomValue * 0.2);
			}
	}

	get genomes() { return this._genomes; }
	get ANNs() { return this._ANNs; }

}

//

export default GeneticAlgorithm;

File: carDemo/src_cpp/src/Experiment/Simulation/ai/GeneticAlgorithm.cpp


#include "GeneticAlgorithm.hpp"


#include "../utils/TraceLogger.hpp"
#include "../utils/randomiser.hpp"


#include <ctime> // <- time


GeneticAlgorithm::GeneticAlgorithm()
	:	m_current_id(1),
		m_current_generation(1),
		m_best_fitness(0.0f),
		m_is_a_great_generation(false)
{
	my_srand( time(NULL) );
}

void	GeneticAlgorithm::init(NeuralNetworkTopology& NNTopology)
{
	m_pNNTopology = &NNTopology;
	this->generateRandomPopulation();
}

void	GeneticAlgorithm::generateRandomPopulation()
{
	// reset the genomes
	m_genomes.resize(200);

	for (unsigned int i = 0; i < m_genomes.size(); ++i)
	{
		t_genome& genome = m_genomes[i];

		genome.m_index = i;
		genome.m_id = m_current_id++;
		genome.m_fitness = 0.0f;

		genome.m_weights.resize( m_pNNTopology->getTotalWeights() );

		for (float& weight : genome.m_weights)
			weight = randomClamped();
	}




	m_NNetworks.reserve( m_genomes.size() );
	for (unsigned int i = 0; i < m_genomes.size(); ++i)
	{
		m_NNetworks.push_back( NeuralNetwork(*m_pNNTopology) );

		m_NNetworks[i].setWeights( m_genomes[i].m_weights );
	}





	m_current_generation = 1;
	m_best_fitness = 0.0f;

	m_alpha_genome.m_index = -1;
	m_alpha_genome.m_id = -1;
	m_alpha_genome.m_fitness = 0.0f;
}

void	GeneticAlgorithm::BreedPopulation()
{
	if (m_genomes.empty())
		return;

	std::vector<t_genome>	bestGenomes;
	getBestGenomes(20, bestGenomes);


	std::vector<t_genome>	children;
	children.reserve( m_genomes.size() );
	

	if (m_best_fitness < bestGenomes[0].m_fitness)
	{
		m_best_fitness = bestGenomes[0].m_fitness;

		D_MYLOG("m_current_generation=" << m_current_generation
			<< std::fixed << ", m_best_fitness=" << m_best_fitness);

		m_is_a_great_generation = true;
	}
	else
	{
		m_is_a_great_generation = false;
	}


	unsigned int current_index = 0;

	if (m_best_fitness > m_alpha_genome.m_fitness)
	{
		m_alpha_genome = bestGenomes[0];
	}
	else
	{
		// elitism :
		// here the best (alpha) genome from previous
		// generation is pushed as an unmodified child

		t_genome	bestDude;
		bestDude.m_fitness = 0.0f;
		bestDude.m_id = m_alpha_genome.m_id;
		bestDude.m_weights = m_alpha_genome.m_weights;
		bestDude.m_index = current_index++;

		// mutate(bestDude);
		children.push_back(bestDude);
	}


	{
		// Carry on the best dude.
		t_genome	bestDude;
		bestDude.m_fitness = 0.0f;
		bestDude.m_id = bestGenomes[0].m_id;
		bestDude.m_weights = bestGenomes[0].m_weights;
		bestDude.m_index = current_index++;

		mutate(bestDude);
		children.push_back(bestDude);
	}

	// breed best genomes

	for (unsigned int i = 0; i < bestGenomes.size(); ++i)
		for (unsigned int j = i+1; j < bestGenomes.size(); ++j)
		{
			t_genome	baby1, baby2;

			CrossBreed(bestGenomes[i], bestGenomes[j], baby1, baby2);
			mutate(baby1);
			mutate(baby2);

			baby1.m_index = current_index++;
			baby2.m_index = current_index++;

			if (children.size() < children.capacity())
				children.push_back(baby1);
			if (children.size() < children.capacity())
				children.push_back(baby2);
		}

	// For the remainding n population, add some random kiddies.
	int remainingChildren = static_cast<int>(m_genomes.size() - children.size());

	for (int i = 0; i < remainingChildren; i++)
	{
		t_genome	genome;
		genome.m_id = m_current_id++;
		genome.m_fitness = 0.0f;

		genome.m_weights.resize( m_pNNTopology->getTotalWeights() );

		for (float& weight : genome.m_weights)
			weight = randomClamped();

		genome.m_index = current_index++;
		children.push_back( genome );
	}

	m_genomes = children;
	++m_current_generation;

	for (unsigned int i = 0; i < m_genomes.size(); ++i)
		m_NNetworks[i].setWeights( m_genomes[i].m_weights );
}

void	GeneticAlgorithm::getBestGenomes(unsigned int totalAsked, std::vector<t_genome>& out) const
{
	// realistic total outputed genomes 
	totalAsked = std::min<unsigned int>(m_genomes.size(), totalAsked);

	out.clear();

	while (out.size() < totalAsked)
	{
		float bestFitness = 0;
		int bestIndex = -1;
		for (unsigned int i = 0; i < m_genomes.size(); i++)
			if (m_genomes[i].m_fitness > bestFitness)
			{
				bool isUsed = false;

				// no duplicate
				for (unsigned int j = 0; j < out.size(); j++)
					if (out[j].m_id == m_genomes[i].m_id)
					{
						isUsed = true;
						break;
					}

				if (isUsed)
					continue;

				bestIndex = i;
				bestFitness = m_genomes[bestIndex].m_fitness;
			}

		if (bestIndex != -1)
			out.push_back( m_genomes[bestIndex] );
	}
}

void	GeneticAlgorithm::CrossBreed(const t_genome& g1, const t_genome& g2, t_genome& baby1, t_genome& baby2)
{
	unsigned int totalWeights = g1.m_weights.size();
	unsigned int crossover = std::min(randomFloat(), 0.99f) * totalWeights;

	baby1.m_id = m_current_id++;
	baby1.m_weights.resize(totalWeights);

	baby2.m_id = m_current_id++;
	baby2.m_weights.resize(totalWeights);

	for (unsigned int i = 0; i < crossover; i++)
	{
		baby1.m_weights[i] = g1.m_weights[i];
		baby2.m_weights[i] = g2.m_weights[i];
	}

	for (unsigned int i = crossover; i < totalWeights; i++)
	{
		baby1.m_weights[i] = g2.m_weights[i];
		baby2.m_weights[i] = g1.m_weights[i];
	}
}

void	GeneticAlgorithm::mutate(t_genome& genome) const
{
	int total_mutated = 0;

	for (float& weight : genome.m_weights)
		if (randomFloat() < 0.1f)
		{
			weight += (randomClamped() * 0.2f);

			++total_mutated;
		}
}

